### PR TITLE
feat: Add new read filtering to datasources guardian

### PR DESCRIPTION
**Pull request by aarongodin**
This adds a new function to the datasources guardian for separating access control to datasources by each of "read" and "query" operations. The change will allow us to provide varying access checks for datasources API calls to remediate an issue of inconsistent behavior between `datasources:read` and `datasources:query` actions.

### CODE REVIEW

File: pkg/api/datasources.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ func (hs *HTTPServer) GetDataSources(c *contextmodel.ReqContext) response.Respon
 		return response.Error(http.StatusInternalServerError, "Failed to query datasources", err)
 	}
 
-	filtered, err := hs.dsGuardian.New(c.SignedInUser.OrgID, c.SignedInUser).FilterDatasourcesByQueryPermissions(dataSources)
+	filtered, err := hs.dsGuardian.New(c.SignedInUser.OrgID, c.SignedInUser).FilterDatasourcesByReadPermissions(dataSources)
```
Comment (IevaVasiljeva) [inline pkg/api/datasources.go:57]: I wonder if we need to make the same change [here](https://github.com/grafana/grafana/blob/main/pkg/api/frontendsettings.go#L402)?

**Review by IevaVasiljeva — APPROVED** [review]
LGTM with one suggestion.

File: pkg/api/datasources.go (hunk 1 above)
Comment (aarongodin) [reply pkg/api/datasources.go:57]: thanks, I remember seeing that but forgot to update it. https://github.com/grafana/grafana/pull/91345/commits/09999744333495bd19dbc82600300e026ec7db27

**Comment by eleijonmarck** [comment]
pushing main, as we were quite behind.

**Review by eleijonmarck — APPROVED** [review]
I like this approach! additional filtering just makes sense

**Pull request merged**